static_src/assignment_9.md

```diff
@@ -0,0 +1,53 @@
+
+## HTTP gzip compression
+
+When our webserver receives a request with header "Accept-Encoding: gzip, deflate", we will generate a response with the appropriate RequestHandler, and then call our GzipHandler's HandleRequest method. The GzipHandler will extract the response's body, compress it, update the response's Content-Length, and add the "Content-Encoding: gzip" header.
+
+For example, using Firefox's network browser tools to track responses returned from our webserver, one can see the "Content-Encoding: gzip" header included when a request is sent for ec2-35-166-129-63.us-west-2.compute.amazonaws.com:8080/static/PMT.gif
+
+We used boost's GzipFilters (http://www.boost.org/doc/libs/1_39_0/libs/iostreams/doc/classes/gzip.html). Furthermore, we used gzip.cpp (http://www.boost.org/doc/libs/1_62_0/libs/iostreams/src/gzip.cpp) and zlib.cpp (http://www.boost.org/doc/libs/1_62_0/libs/iostreams/src/zlib.cpp), both created by Jonathan Turkanis.
+
+#### HTTP compression design process
+
+HTTP compression had a bit more difficult design decisions than Markdown rendering.
+
+First, we needed to once again pick a library for use in our program. This was a fairly easy choice, as boost provides functionality for Gzip, and these boost functions could be plugged in to the code fairly easy once certain linker dependencies and header file concerns were resolved. We looked at other library options briefly, but this was the clear cut most simple and efficient choice for us.
+
+Additionally, we needed to decide how compression should be handled (i.e. where in the code should compression occur & what object handles the compression). We initially placed the compression code in the same handler that handles serving static files. We quickly realized that this was not a sustainable decision, as compression may need to be utilized for other types of requests and because we may want to turn off compression for static files depending on various factors. Our decision then was to create a handler specifically for Gzip compression, that would work on a response created by another handler. The only API change needed was to add a method to read the body of a Response object, which is not going to cause any issue in working with the code in the future. This allows for configuration of the server as well as the ability to compress our responses for different types of requests. The flexibility for future design was key in making this decision.
```

Comment:
actually, the server automatically does gzip. its not configured in the configuration file